crates/indexer/src/lib.rs

```diff
@@ -0,0 +1,816 @@
+            logs.len(),
+            from_block,
+            to_block
+        );
+
+        for (log, log_data) in logs {
+            tracing::debug!("Processing callback failed event for request: 0x{:x}", log.requestId);
+            let tx_hash = log_data.transaction_hash.context("Transaction hash not found")?;
+            let (metadata, _tx_input) = match cache.get(tx_hash) {
+                Some((metadata, tx_input)) => (metadata.clone(), tx_input),
+                None => {
+                    let tx = self
+                        .boundless_market
+                        .instance()
+                        .provider()
+                        .get_transaction_by_hash(tx_hash)
+                        .await?
+                        .context(anyhow!(
+                            "Transaction not found for hash: {}",
+                            hex::encode(tx_hash)
+                        ))?;
+
+                    let block_number = tx.block_number.context("block number not found")?;
+                    let block_timestamp = match log_data.block_timestamp {
+                        Some(ts) => ts,
+                        None => self.block_timestamp(block_number).await?,
+                    };
+                    let metadata =
+                        TxMetadata::new(tx_hash, tx.from(), block_number, block_timestamp);
+                    cache.insert(metadata.clone(), tx.input().clone());
+                    (metadata, &tx.input().clone())
+                }
+            };
+
+            self.db
+                .add_callback_failed_event(
+                    log.requestId,
+                    log.callback,
+                    log.error.to_vec(),
+                    &metadata,
+                )
+                .await?;
+        }
+
+        Ok(())
+    }
+
+    async fn current_block(&self) -> Result<u64, ServiceError> {
+        Ok(self.boundless_market.instance().provider().get_block_number().await?)
+    }
+
+    async fn block_timestamp(&self, block_number: u64) -> Result<u64, ServiceError> {
+        Ok(self
+            .boundless_market
+            .instance()
+            .provider()
+            .get_block_by_number(BlockNumberOrTag::Number(block_number))
+            .await?
+            .context(anyhow!("Failed to get block by number: {}", block_number))?
+            .header
+            .timestamp)
+    }
+}
+
+// Cache for transactions to avoid multiple calls to the provider
+struct TxCache {
+    // Mapping from transaction hash to (from, block number, block timestamp, tx input)
+    cache: HashMap<B256, (TxMetadata, Bytes)>,
+}
+
+impl TxCache {
+    fn new() -> Self {
+        Self { cache: HashMap::new() }
+    }
+
+    fn get(&self, tx_hash: B256) -> Option<&(TxMetadata, Bytes)> {
+        self.cache.get(&tx_hash)
+    }
+
+    fn insert(&mut self, metadata: TxMetadata, tx_input: Bytes) {
+        self.cache.insert(metadata.tx_hash, (metadata, tx_input));
+    }
+}
```

Comment:
We should clear this cache periodically just to be safe

Reply:
Absolutely, that is cleared every time `process_blocks` is called